### ISSUE TITLE
Custom date picker for date fields in AMP forms

### ISSUE THREAD

**Issue by rudygalfi** [ampproject/amphtml#6469]
This issue is opened to collect requirements and track development of being able to use a custom date picker in conjunction with the date field support in AMP forms.

**Comment by ericlindley-g** [comment]
/cc @dandv @choumx @aghassemi 

FYI
This has been coming up more recently, now that a lot of the needs for e-commerce have been addressed. Chatted w/ Malte about the feature, and current thinking is that we may be able to take an approach similar to autocomplete, where we both 

a) expose the events in bind to make this work and create an ABE that demos how folks could compose an advanced case, and 
b) create a simple, good-UX solution encapsulated as a higher-level component, as well

**Comment by vsiv** [comment]
Can't wait for autocomplete and this. 👍

**Comment by dandv** [comment]
+1. Date *range* picking has been requested by travel partners - searching for hotel availability for example.

**Comment by aghassemi** [comment]
I highly recommend we try to find an existing good open-source implementation and integrate it. A high quality, mobile friendly date picker with a11y and i18n support is a lot of work. Supporting non-Gregorian calendars adds a lot to it as well.

Looking at Airbnb's version (https://github.com/airbnb/react-dates): 500 commits, 57 contributors, thousands lines of code...

**Comment by aghassemi** [comment]
**UX Design Review**

- [ ] Allow typing (new version of ReactDate does). 
- [ ] @spacedino to come up with a base style for calendar (Decision is to have a "good" (instead of a "reset" that picks up browser default) base style)
- [ ] @cvializ to "reset" the `input` styles back to browser defaults.
- [ ] Modal for modal overlay for desktop. 
- [ ] @cvializ publish Demo link for @spacedino to use for styling.

**Comment by spacedino** [comment]
Final designs and spex. The final design we decided on was not actually a11y color compliant so I had to change. Open to discuss if there are issues but this should work for now. We might want to make it more obvious which are the start and end dates by changing the shape since the color is pretty dark now. Or we could decide that having the start and end dates as a dark color, but the middle dates with a lighter bg and black text is ok too (19). 

FINAL

![date_picker_final_empty_16](https://user-images.githubusercontent.com/7339798/34040988-3e3e4a42-e1ea-11e7-83b4-ea291a4dfcf4.png)
![date_picker_final_full_17](https://user-images.githubusercontent.com/7339798/34040989-3e6d7bbe-e1ea-11e7-9cd8-dba122aa2994.png) OR
![date_picker_final_full_19](https://user-images.githubusercontent.com/7339798/34040990-3e9dad8e-e1ea-11e7-9113-4fc081b21085.png)
![date_picker_single_18](https://user-images.githubusercontent.com/7339798/34040991-3ed51df0-e1ea-11e7-974e-e250a3fbb5a5.png)
![final_spex_date_picker](https://user-images.githubusercontent.com/7339798/34040992-3f047992-e1ea-11e7-8def-6770d9fbeb1b.png)


/cc +aghassemi +cvisliz +lswang1618

**Comment by spacedino** [comment]
from @cvializ demos : 

- https://ampcvializ.herokuapp.com/examples/date-picker.amp.html
- [ AMP.toggleExperiment('amp-date-picker', true) to view ]

**Next steps:**
- launch in experimental mode
- update visual designs
- final UX review

**Comment by lswang1618** [comment]
This component is ready for experimenting. In lieu of full documentation, here are some details for experimentation. The component is called amp-date-picker with type=range or type=single for selecting a single date or a range of dates. 

A placeholder is required for the component - for a single date, this can take the form:  
`<input amp-date-placeholder placeholder="Pick a date" id="date"/>`

For a date range:
```
<input amp-date-placeholder-start placeholder="Start date" id="start" />
<input amp-date-placeholder-end placeholder="End date" id="end" />
```

The properties available for both include: 
- max: the latest date that the calendar will allow. The default is unlimited.
- min: the earliest date that the calendar will allow. The default is the current date.
- src: a link to a JSON file to configure the date picker. 
- locale: a locale string, e.g. "en", "fa", "zh".
- format: a standard date string e.g. YYYY-MM-DD to use as the output for AMP Events. This will be formatted in the output locale, unless the output locale is Unix seconds or millis i.e. "X" or "x"
- blocked: a space separated list of ISO 8601 dates and RFC 5545 RRULEs specifying disallowed dates. Does not require amp-mustache.
- highlighted: a space separated list of ISO 8601 dates and RFC 5545 RRULEs specifying dates displayed with a highlight style. Does not require amp-mustache.

Properties that are inherited from the Airbnb library (https://github.com/airbnb/react-dates) include:

**For both range and single**
- anchor-direction
- auto-focus
- disabled
- enable-outside-days
- first-day-of-week
- horizontal-margin
- initial-visible-month
- keep-open-on-date-select
- minimum-nights
- month-format
- number-of-months
- orientation
- reopen-picker-on-clear-date
- required
- screen-reader-input-message
- show-default-input-icon
- with-full-screen-portal
- with-portal

**For single**
- initial-date
- show-clear-date

**For range**
- auto-focus-end-date
- initial-end-date
- initial-start-date
- show-clear-dates

You can also use amp-mustache to create templates for custom date markup and an info-panel below the calendar. Other features include tooltips, setDate(s) and clear actions, and change event listeners. Examples of these implementations can be found here: 
- https://github.com/ampproject/amphtml/blob/master/examples/date-picker.amp.html
- https://github.com/ampproject/amphtml/blob/master/examples/travel.amp.html

**Comment by eric30** [comment]
I tried to follow all the steps and gave it a try, however it didn't work out. There's a 404 error occurred:

![image](https://user-images.githubusercontent.com/1381195/34368920-cf5659be-eaf2-11e7-905f-aada64d3eb8a.png)

Is this a bug or did I do something wrong?

**Comment by lswang1618** [comment]
Update that this is now working in the Dev Channel.

**Comment by adnanyousafch** [comment]
Any timeline for when it will be out of experimental stage?

**Comment by aghassemi** [comment]
@lswang1618 @cvializ Given default styles are now in, I personally feel comfortable moving this to at least a doc-level optin so folks can start shipping in 3-4 weeks,  thoughts? (@cvializ I guess we still need the validation rules).

**Comment by lswang1618** [comment]
Sure, sgtm.

**Comment by vonkow** [comment]
Any update on this timeline?

**Comment by mdiblasio** [comment]
Should the calendars be hidden by default or do they need to be implemented inside of a lightbox?

[This demo](https://ampcvializ.herokuapp.com/examples/date-picker.amp.html) has the calendars hidden, but [AMPByExample Demo](https://ampbyexample.com/components/amp-date-picker/) displays them by default. 

Which is expected?

**Comment by rudygalfi** [comment]
@mdiblasio Isn't it a user experience choice for the publisher? cc @lswang1618

**Comment by mdiblasio** [comment]
I'm more referring to the inconsistency between the 2 demos, which have nearly the same implementations.

Demo 1:
![demo 1 ](https://user-images.githubusercontent.com/17748781/37980607-7e4f1d5e-31b9-11e8-9529-8ef6da199e2e.png)

Demo 2:
 ![demo 2](https://user-images.githubusercontent.com/17748781/37980584-6e918384-31b9-11e8-9475-ad4c05d62cca.png)

Neither use a lightbox but the first demo doesn't display the calendar until the field is clicked. If the publisher doesn't want to show the calendar until clicked, is there an attribute or must they use a lightbox?

**Comment by lswang1618** [comment]
It's an attribute - use mode:overlay

**Comment by ampprojectbot** [comment]
This issue hasn't been  updated in awhile. @lswang1618 Do you have any updates?

**Comment by 88kbbq** [comment]
What about validation? How do you make this date picker a mandatory field like e-mail or name? Documentation for validation seems not to apply to this component.

**Comment by 88kbbq** [comment]
> What about validation? How do you make this date picker a mandatory field like e-mail or name? Documentation for validation seems not to apply to this component.

Ah, the `required` attribute needs to appear in the input field, not the amp-date-picker component. Tricky!

**Comment by aghassemi** [comment]
has been implemented and launched. If you have any FR or bugs, please fine a new ticket.

**Comment by Pachara** [comment]
I am looking for Time Picker.
Any Ideas?
